Add slog handlers to match existing logrus formatters

rosstimothy opened the pull request:
Implements two new `slog.Handlers` that attempt to output text in the same format as existing `logrus.TextFormatter` implementations. The most notable differences are that the order of fields in the json output does NOT match the logrus order and the additional items in text output are NOT sorted alphabetically like they are in logrus.

Tests added validate the output is equivalent when using slog or logrus taking into account the differences noted above . A benchmark test is also included to demonstrate the improvements slog offers over logrus. Some of the improvements made for the slog handlers were also able to be applied to the logrus formatters as well.

strideynet commented on lib/utils/log/slog_handler.go:
```diff
@@ -0,0 +1,386 @@
+// SlogJSONHandler is a [slog.Handler] that outputs messages in a json
+// format per the config file.
+type SlogJSONHandler struct {
+	handler *slog.JSONHandler
```
Super nit/musing: We could embed this here to avoid the need to write out the methods that delegate to `handler`.

strideynet commented on lib/utils/log/slog_handler.go:
```diff
@@ -0,0 +1,386 @@
+	file = s.File[idx+1:]
+	line = s.Line
+
+	return
```
Nit: It doesn't really matter but I really dislike naked returns - I find em super hard to read and they're fairly rare across our codebase.

```suggestion
	return file, line
```

strideynet commented on lib/utils/log/formatter_test.go:
```diff
@@ -0,0 +1,381 @@
+				formatter.timestampEnabled = true
+				require.NoError(t, formatter.CheckAndSetDefaults())
+
+				logger := logrus.New()
```
Nit: I'd probably call this `logrusLogger` or similar - just to distinguish it clearly from `slogLogger`.

strideynet commented on lib/utils/log/slog_handler.go:
```diff
@@ -0,0 +1,386 @@
+	component      string
+	preformatted   []byte   // data from WithGroup and WithAttrs
+	unopenedGroups []string // groups from WithGroup that haven't been opened
+	mu             *sync.Mutex
```
We should add a comment to `mu` to state this is protecting `out` and not the rest of the fields and that it's intentional that mu is copied to child handlers. IIRC there was actually a bug in the original exp/slog implementation where a new mutex was created during the creation of child handlers and this lead to garbled logs - whilst we've not repeated that bug here - I think it's trippy enough that a sentence or two is justified.

strideynet reviewed (approved):
Looks good - nice work - I can't imagine it was easy getting these formats to match 😓 

I've raised a couple thoughts - mostly nits and a few requests to add some more comments. Since the logger needs to focus on performance, the code is going to be a bit less readable than most - so any comments we can add for the next person to stumble thru will be a blessing.

rosstimothy committed:
Add slog handlers to match existing logrus formatters

Implements two new slog.Handlers that attempt to output text in the
same format as existing logrus.TextFormatter implementations. The
most notable differences are that the order of fields in the json
output does NOT match the logrus order and the additional items
in text output are NOT sorted alphabetically like they are in logrus.

Tests added validate the output roughly matches between the two
loggers given the differences noted above. A benchmark test is also
included to demonstrate the improvements slog offers over logrus.
